File: notes/0399/0399.py
class Solution:    
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        def dfs(x, y, visited):
            if x == y:
                return 1.0
            visited.add(x)
            for n in graph[x]:
                if n in visited:
                    continue
                visited.add(n)
                d = dfs(n, y, visited)
                if d > 0:
                    return d * graph[x][n]
            return - 1.0
        
        graph = collections.defaultdict(dict)
        for (x, y), v in zip(equations, values):
            graph[x][y] = v 
            graph[y][x] = 1.0 / v
        # res = []
        # for x, y in queries:
        #     if x not in graph and y not in graph:
        #         res.append(-1.0)
        #     else:
        #         res.append(dfs(x, y, set()))
        # return res
        return [dfs(x, y, set()) if x in graph and y in graph else -1.0 for x, y in queries]
        

File: notes/0399/0399.cpp
class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, unordered_map<string, double>> graph;
        const int size = equations.size();
        vector<double> ans;
        
        for (int i=0; i<size; i++) {
            string first = equations[i][0];
            string second = equations[i][1];
            graph[first][second] = 1.0*values[i];
            graph[second][first] = 1.0/values[i];
        }
        
        for (auto p: queries) {
            if (!graph.count(p[0]) || !graph.count(p[1])) {
                ans.push_back(-1.0);
                continue;
            }
            unordered_set<string> visited;
            ans.push_back(dfs(graph, visited, p[0], p[1]));
        }
        return ans;
    }
private:
    double dfs(unordered_map<string, unordered_map<string, double>>& graph, unordered_set<string>& visited, string i, string j) {
        if (i==j) return 1.0;
        for (auto p: graph[i]) {
            string tmp = p.first;
            double tmp_val = p.second;
            if (visited.count(tmp)) continue;
            visited.insert(tmp);
            double val = dfs(graph, visited, tmp, j);
            if (val != -1.0) return val*tmp_val;
        }
        return -1.0;
    } 
};
